File: ex67.py
class Solution:
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        r = ''

        la = list(str(a))
        lb = list(str(b))

        #Keep two binaries at the same length
        if len(la) > len(lb):
            for i in range(len(la) - len(lb)):
                lb.insert(0, '0')
        elif len(lb) > len(la):
            for i in range(len(lb) - len(la)):
                la.insert(0, '0')

        #Add
        cache = 0

        for i in range(len(la) - 1, -1, -1):
            s = int(la[i]) + int(lb[i])

            if s + cache == 0:
                r = '0' + r
                cache = 0
            elif s + cache == 1:
                r = '1' + r
                cache = 0
            elif s + cache == 2:
                r = '0' + r
                cache = 1
            elif s + cache == 3:
                r = '1' + r
                cache = 1

        #If cache is left, add 1 to the front
        if cache == 1:
            r = '1' + r

        return r


sol = Solution()
print(sol.addBinary(1010, 1011))

File: code/ex67.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        int n1 = a.length(), n2 = b.length();
        
        int p1 = n1 - 1, p2 = n2 - 1, carry = 0;
        string res;
        while (p1 >= 0 || p2 >= 0) {
            int d1 = p1 >= 0 ? a[p1--] - '0' : 0;
            int d2 = p2 >= 0 ? b[p2--] - '0' : 0;
            res = to_string((d1 + d2 + carry) % 2) + res;
            carry = (d1 + d2 + carry) / 2;
        }
        if (carry > 0)    res = '1' + res;
        
        return res;
    }
};